website/mailing_list/utils.py

```diff
@@ -0,0 +1,422 @@
+    """ Adds/updates single member of a mailing list on Mailgun
+    Called to add, subscribe, or unsubscribe a user.
+
+    :param str node: ID of node in question
+    :param str user: ID of user to update
+    :param str email: email address to add. If None, `user.username` assumed.
+    :param bool enabled: Enable or disable user?
+    :param str old_email: Previous email of this user in list, included when user changes primary email.
+    """
+    from website.models import Node, User  # avoid circular import
+    node = Node.load(node_id)
+    user = User.load(user_id)
+    email = email or user.username
+
+    if old_email:
+        resp = requests.put(
+            '{}/{}/members/{}'.format(MAILGUN_BASE_LISTS_URL, address(node_id), old_email),
+            auth=('api', settings.MAILGUN_API_KEY),
+            data={
+                'subscribed': 'no',
+                'vars': json.dumps({'_id': user_id, 'primary': False})
+            }
+        )
+        if resp.status_code not in [200, 404]:
+            raise HTTPError(resp.status_code, data={'message_long': resp.json() or ''})
+
+    resp = requests.post(
+        '{}/{}/members'.format(MAILGUN_BASE_LISTS_URL, address(node._id)),
+        auth=('api', settings.MAILGUN_API_KEY),
+        data={
+            'address': email,
+            'subscribed': enabled and email == user.username and user not in get_unsubscribes(node),
+            'vars': json.dumps({'_id': user._id, 'primary': email == user.username or bool(old_email)}),
+            'upsert': True
+        }
+    )
+    if resp.status_code != 200:
+        raise HTTPError(resp.status_code, data={'message_long': resp.json() or ''})
+
+@require_mailgun
+def remove_user_from_list(node_id, user_id):
+    """ Removes single member of a mailing list on Mailgun
+    Called when a contributor is removed from a Node.
+
+    :param str node_id: ID of node in question
+    :param str user_id: ID of user to remove
+    """
+    from website.models import User  # avoid circular import
+    user = User.load(user_id)
+
+    for email in user.emails:
+        resp = requests.delete(
+            '{}/{}/members/{}'.format(MAILGUN_BASE_LISTS_URL, address(node_id), email),
+            auth=('api', settings.MAILGUN_API_KEY)
+        )
+        if resp.status_code not in [200, 404]:
+            raise HTTPError(resp.status_code, data={'message_long': resp.json() or ''})
+
+@require_mailgun
+def update_multiple_users_in_list(node_id, members):
+    """ Adds/updates members of a mailing list on Mailgun
+
+    :param str node_id: The id of the node in question
+    :param list members: List of json-formatted user dicts to add/enable/update
+    """
+    resp = requests.post(
+        '{}/{}/members.json'.format(MAILGUN_BASE_LISTS_URL, address(node_id)),
+        auth=('api', settings.MAILGUN_API_KEY),
+        data={
+            'upsert': True,
+            'members': json.dumps(members)
```

Comment:
According to the mailgun API, there is a limit of 1,000 members per request.

https://documentation.mailgun.com/api-mailinglists.html#mailing-lists
> <POST /lists/<address>/members.json
>> Adds multiple members, up to 1,000 per call, to a Mailing List.

For large projects, this could be problematic- especially once we start offering `project-all` lists that include child components.

Suggested improvement: all calls to the mailgun API should plan for rate limiting behavior. We are planning several large reproducibility initiatives, as well as targeting larger institutions... so this edge case is not too far-fetched.